### PR TITLE
Symphonia codec opus

### PR DESCRIPTION
**Pull request by jiztastamablastamarang**
## Description:
This PR is part of my [capstone project](https://github.com/rust-lang-ua/rustcamp/tree/master/6_project) for the [Ukrainian Rust Community](https://uarust.com/)  Bootcamp. It introduces a new [Opus decoder](https://datatracker.ietf.org/doc/html/rfc6716) implementation as part of the [Symphonia](https://docs.rs/symphonia/latest/symphonia/) crate and addresses [#8](https://github.com/pdeljanov/Symphonia/issues/8). The decoder currently  includes foundational implementations for entropy decoding, packet parsing and supports the **SILK** mode prototype decoder. CELT and hybrid modes are reserved for future implementation.

## Key Changes:
1. **OpusDecoder Implementation**:
   - Added the `OpusDecoder` struct that implements the `Decoder` trait.
   - Supports **SILK mode** decoding.
   - Includes methods for decoding packets, resetting the decoder state, and retrieving codec parameters.
   - Introduced a codec descriptor `OPUS_CODEC_DESCRIPTOR` using `Lazy` to instantiate the decoder within Symphonia.

2. **Entropy Decoding**:
   - Implemented the `RangeDecoder` trait, which handles the entropy coding based on r[ange coding](https://datatracker.ietf.org/doc/html/rfc6716#section-4.1).
   - Provides methods for decoding symbols using inverse cumulative distribution functions (ICDF) and logarithmic probabilities.
   - Supports handling raw bits for CELT layers in the Opus format, improving stream resilience to bit errors.

3. **Header Parsing**:
   - Implemented parsing for Opus headers according to [RFC 7845](https://datatracker.ietf.org/doc/html/rfc7845).
   - Added structures to parse and validate the **Identification Header** and **Comment Header**.
   - Provides detailed error handling for invalid header parameters such as channel count, input sample rate, and output gain.

4. **Opus Packet Handling**:
   - Introduced functionality to parse Opus packets, supporting various configurations, including frame counting and padding handling.
   - Implemented **Code 0**, **Code 1**, **Code 2**, and **Code 3** packet parsing.
   - Added a method to handle both constant bit rate (CBR) and variable bit rate (VBR) packets, including error checks for invalid frame sizes and total audio duration.

### Tests:
- Unit tests for:
  - Opus header parsing.
  - Entropy decoding (ICDF and logarithmic probabilities).
  - Packet parsing for all packet types (Code 0, Code 1, Code 2, Code 3).
  - Decoder functionality including frame decoding and parameter retrieval.

### Future Work:
- Implement support for **CELT** and **Hybrid** decoding modes.
- Integrate full functionality for **finalizing** decoders and handling last decoded frames.
- Optimize entropy decoding for performance and support for CELT raw bit handling.

### CODE REVIEW

File: symphonia-codec-opus/Cargo.toml [hunk 1]
```diff
@@ -14,4 +14,19 @@ rust-version = "1.53"
 
 [dependencies]
 log = "0.4"
-symphonia-core = { version = "0.5.4", path = "../symphonia-core" }
\ No newline at end of file
+symphonia-core = { version = "0.5.4", path = "../symphonia-core" }
+once_cell = "1.19.0"
```
Comment (NYBACHOK) [inline symphonia-codec-opus/Cargo.toml]: Why not use std?

File: symphonia-codec-opus/Cargo.toml [hunk 2]
```diff
@@ -14,4 +14,19 @@ rust-version = "1.53"
...
+lazy_static = "1.5.0"
```
Comment (NYBACHOK) [inline symphonia-codec-opus/Cargo.toml]: Use `lazy_static` only in `no_std` enviroment.

File: symphonia-codec-opus/src/silk/decoder.rs [hunk 3]
```diff
@@ -0,0 +1,1102 @@
...
+        let mut ltp_signal = vec![0.0; excitation.len()];
+
+        for i in 0..excitation.len() {
+            let mut pred = 0.0;
+            for (j, &coeff) in ltp_coeffs.iter().enumerate() {
+                if i >= pitch_lag as usize + j {
+                    pred += ltp_signal[i - pitch_lag as usize - j] * (coeff as f32 / 128.0);
+                }
+            }
+            ltp_signal[i] = excitation[i] + ltp_scale * pred;
```
Comment (NYBACHOK) [inline symphonia-codec-opus/src/silk/decoder.rs]: I hope you checked that no panic may occur

**Review by NYBACHOK — APPROVED** [review]
Last thing. Consider usage of `AsRef<[u8]>` for public methods

**Review by FairlyBet — APPROVED** [review]

File: symphonia-codec-opus/Cargo.toml (hunk 1 above)
Comment (jiztastamablastamarang) [reply symphonia-codec-opus/Cargo.toml]: Fixed 3c92322

File: symphonia-codec-opus/src/silk/decoder.rs (hunk 3 above)
Comment (jiztastamablastamarang) [reply symphonia-codec-opus/src/silk/decoder.rs]: This module may panic during development. I will debug, add tests, and replace direct indexing with safe access after refinement, as done in other modules (e.g., entropy).

File: symphonia-codec-opus/Cargo.toml (hunk 2 above)
Comment (jiztastamablastamarang) [reply symphonia-codec-opus/Cargo.toml]: Removed 3c92322 It was unused dep

File: symphonia-codec-opus/Cargo.toml (hunk 1 above)
Comment (jiztastamablastamarang) [reply symphonia-codec-opus/Cargo.toml]: Fixed 3c92322

**Review by ikripaka — APPROVED** [review]
By now I've dealt with local codec written for audio compression, bit it doesn't compare with this. It looks insane!

**Review by Jujumba — APPROVED** [review]

**Review by VladyslavNesterovskyi — APPROVED** [review]

**Review by TheBestTvarynka — APPROVED** [review]

**Review by totalsnack — APPROVED** [review]

**Pull request merged**